WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/eclipse/logic/event/listeners/DebugEventListener.java

```diff
@@ -0,0 +1,119 @@
+package nl.tudelft.watchdog.eclipse.logic.event.listeners;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+import org.eclipse.debug.core.DebugEvent;
+import org.eclipse.debug.core.IDebugEventSetListener;
+import org.eclipse.debug.core.model.IThread;
+import org.eclipse.debug.core.model.IVariable;
+import org.eclipse.debug.core.model.IWatchExpression;
+
+import nl.tudelft.watchdog.core.logic.event.EventManager;
+import nl.tudelft.watchdog.core.logic.event.eventtypes.DebugEventBase;
+import nl.tudelft.watchdog.core.logic.event.eventtypes.EventType;
+
+/**
+ * Class that handles all {@link DebugEvent}s and generates the appropriate
+ * instances of {@link DebugEventBase} which are then passed to the
+ * {@link EventManager}.
+ */
+public class DebugEventListener implements IDebugEventSetListener {
+
+	/**
+	 * The {@link EventManager} used for persisting and transferring the debug
+	 * events.
+	 */
+	private final EventManager eventManager;
+
+	/**
+	 * List with the hashes of all current watch expressions. Used to avoid
+	 * duplicate 'Define Watch' events.
+	 */
+	private List<Integer> watchExpressionHashes;
+
+	/** Constructor. */
+	public DebugEventListener(EventManager eventManager) {
+		this.eventManager = eventManager;
+		this.watchExpressionHashes = new ArrayList<>();
+	}
+
+	@Override
+	public void handleDebugEvents(DebugEvent[] events) {
+		for (DebugEvent event : events) {
+			if (event.getSource() instanceof IThread) {
+				handleDebugEvent(event);
+			} else if (event.getSource() instanceof IWatchExpression) {
+				handleWatchExpressionEvent(event);
+			} else if (event.getSource() instanceof IVariable
```

Comment:
Extract if conditions into named structures (variables or functions), like isThread, ... ? (Especially the last one is difficult to comprehend)